### ISSUE TITLE
Is it possible to use imenu-anywhere as backend for xref?

### ISSUE THREAD

**Issue by minad** [vspinu/imenu-anywhere#43]

**Comment by vspinu** [comment]
Not currently. I personally very much like not putting all lookup algorithms on one key. I don't think it's workable. Imenu-anywhere is generic and works anywhere but it's not very powerful and is limited to open buffers.  Specialized mode's xref is usually more powerfull and works on files but it's mode specific. Orthogonal concerns in some sense. Works well for me having them on separate keys.

If you really think it's a useful feature, PRs are welcome :)

**Comment by minad** [comment]
Ok, many packages make that choice to rather have separate keys. I would prefer things to be a bit more integrated.
I've found a discussion on the emacs mailing list regarding xref, where you argued for xref allowing multiple backends. Do you think this is still a possibility? Or do you see reasons why one wouldn't want that?

I did quickly hack a xref-imenu integration but it doesn't work if there are other xref backends active. So this would probably only work if there is some additional multi backend, which accesses the other configured backends. I've not written such a multi backend yet though.

**Comment by vspinu** [comment]
@minad Sorry I missed your last comment from last year somehow. 

I haven't tracked the development of xref for the past year but from what I can see the support is not there still, so I guess people din't find it useful. In any case, if ref-backend for imenu is the last backend in the list, it should work as expected, admittedly might not be called often as other backends would kick in. 

Did you manage to integrate imenu-anywhere with the xref on your side? If so, could you provide the code?

**Comment by minad** [comment]
> Did you manage to integrate imenu-anywhere with the xref on your side? If so, could you provide the code?

Yes, I think at some point I managed to write a (very limited) xref backend for imenu-anywhere. However nothing in publishable form. I also thought about writing a multi-xref backend, which supports merging, but nothing came out of that. 

I recently stopped using imenu-anywhere in favor of [Consult](https://github.com/minad/consult). I made `consult-imenu`, and `consult-project-imenu`, which is inspired by imenu-anywhere but more restricted. The idea is the same: caching and merging the imenu from different buffers. Generally `consult-imenu` flattens the imenu similar to what flimenu does and enhances it with a few more Consult features (preview, narrowing etc).
